File: packages/shared/src/types/featureRequest.ts
/**
 * Feature Request / Voting System types
 *
 * Global feature request board where users submit ideas,
 * vote, and comment. Admins manage status lifecycle.
 */

import type { FirestoreTimestamp } from "./common.js";

// =============================================================================
// Enums / Union Types
// =============================================================================

export type FeatureRequestStatus =
  | "open"
  | "under_review"
  | "planned"
  | "in_progress"
  | "completed"
  | "declined";

export type FeatureRequestCategory =
  | "improvement"
  | "new_feature"
  | "integration"
  | "bug_fix"
  | "other";

export type FeatureRequestPriority = "low" | "medium" | "high" | "critical";

export type FeatureRequestSortBy = "votes" | "newest" | "oldest";

// =============================================================================
// Core Types
// =============================================================================

export interface FeatureRequest {
  id: string;
  title: string;
  description: string;
  category: FeatureRequestCategory;
  status: FeatureRequestStatus;
  priority: FeatureRequestPriority | null;
  authorId: string;
  authorDisplayName: string;
  voteCount: number;
  commentCount: number;
  adminResponse: string | null;
  adminResponseAuthorName: string | null;
  adminResponseAt: FirestoreTimestamp | null;
  createdAt: FirestoreTimestamp;
  updatedAt: FirestoreTimestamp;
  /** Enriched on the fly per request — true if the current user has voted */
  hasVoted?: boolean;
}

export interface FeatureRequestComment {
  id: string;
  body: string;
  authorId: string;
  authorDisplayName: string;
  isAdmin: boolean;
  createdAt: FirestoreTimestamp;
  updatedAt: FirestoreTimestamp;
}

// =============================================================================
// Input Types
// =============================================================================

export interface CreateFeatureRequestInput {
  title: string;
  description: string;
  category: FeatureRequestCategory;
}

export interface UpdateFeatureRequestStatusInput {
  status: FeatureRequestStatus;
  adminResponse?: string;
}

export interface SetFeatureRequestPriorityInput {
  priority: FeatureRequestPriority | null;
}

export interface CreateFeatureRequestCommentInput {
  body: string;
}

// =============================================================================
// Response Types
// =============================================================================

export interface FeatureRequestListResponse {
  items: FeatureRequest[];
  nextCursor: string | null;
  totalCount?: number;
}

export interface FeatureRequestDetailResponse {
  request: FeatureRequest;
  comments: FeatureRequestComment[];
  commentsNextCursor: string | null;
}

export interface FeatureRequestVoteResponse {
  voted: boolean;
  voteCount: number;
}

File: EquiDuty-Android/app/src/main/java/com/equiduty/domain/model/FeatureRequest.kt
package com.equiduty.domain.model

data class FeatureRequest(
    val id: String,
    val title: String,
    val description: String,
    val category: FeatureRequestCategory,
    val status: FeatureRequestStatus,
    val priority: FeatureRequestPriority?,
    val authorId: String,
    val authorDisplayName: String,
    val voteCount: Int,
    val commentCount: Int,
    val adminResponse: String?,
    val adminResponseAuthorName: String?,
    val adminResponseAt: String?,
    val createdAt: String,
    val updatedAt: String,
    val hasVoted: Boolean
)

data class FeatureRequestComment(
    val id: String,
    val body: String,
    val authorId: String,
    val authorDisplayName: String,
    val isAdmin: Boolean,
    val createdAt: String,
    val updatedAt: String
)

enum class FeatureRequestStatus(val value: String) {
    OPEN("open"),
    UNDER_REVIEW("under_review"),
    PLANNED("planned"),
    IN_PROGRESS("in_progress"),
    COMPLETED("completed"),
    DECLINED("declined");

    companion object {
        fun fromValue(value: String): FeatureRequestStatus =
            entries.find { it.value == value } ?: OPEN
    }
}

enum class FeatureRequestCategory(val value: String) {
    IMPROVEMENT("improvement"),
    NEW_FEATURE("new_feature"),
    INTEGRATION("integration"),
    BUG_FIX("bug_fix"),
    OTHER("other");

    companion object {
        fun fromValue(value: String): FeatureRequestCategory =
            entries.find { it.value == value } ?: OTHER
    }
}

enum class FeatureRequestPriority(val value: String) {
    LOW("low"),
    MEDIUM("medium"),
    HIGH("high"),
    CRITICAL("critical");

    companion object {
        fun fromValue(value: String): FeatureRequestPriority? =
            entries.find { it.value == value }
    }
}

enum class FeatureRequestSortBy(val value: String) {
    VOTES("votes"),
    NEWEST("newest"),
    OLDEST("oldest")
}
